src/serial.rs

```diff
@@ -99,6 +99,33 @@ const DEFAULT_MODEM_CONTROL: u8 = MCR_OUT2_BIT;
 const DEFAULT_MODEM_STATUS: u8 = MSR_DSR_BIT | MSR_CTS_BIT | MSR_DCD_BIT;
 const DEFAULT_SCRATCH: u8 = 0x00;
 
+/// Defines a series of callbacks that are invoked in response to the occurrence of specific
+/// events as part of the serial emulation logic (for example, when the driver reads data). The
+/// methods below can be implemented by a backend that keeps track of such events by incrementing
+/// metrics, logging messages, or any other action.
+// TODO: The events defined below are just some examples for now to validate the approach. If
+// things look good, we can move on to establishing the initial list. It's also worth mentioning
+// the methods can have extra parameters that provide additional information about the event.
+pub trait SerialEvents {
+    /// The driver reads data from the input buffer.
+    fn buffer_read(&mut self);
+    /// The driver successfully writes a byte data to serial output.
+    fn out_write_all(&mut self);
+    /// A successful `flush` operation has been executed on the serial output.
+    fn out_flush(&mut self);
```

Comment:
I wonder if we can be more restrictive here and have `&self` instead of `&mut self` (`Metric` trait from vmm-sys-util seems to be using &self as well).

Reply:
Sure, it's prob simpler to start like that and we can consider alternatives later if needed.